src/irmin-pack/unix/lower.ml

```diff
@@ -0,0 +1,195 @@
+(*
+ * Copyright (c) 2023 Tarides <contact@tarides.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Import
+include Lower_intf
+
+module Make_volume (Io : Io.S) (Errs : Io_errors.S with module Io = Io) = struct
+  module Io = Io
+  module Errs = Errs
+  module Control = Control_file.Volume (Io)
+
+  type t = {
+    mutable readonly : bool;
+    path : string;
+    control : Control_file.Payload.Volume.Latest.t option;
+  }
+
+  type open_error =
+    [ Io.open_error
+    | `Closed
+    | `Double_close
+    | `Corrupted_control_file
+    | `Unknown_major_pack_version of string ]
+
+  let v ~readonly volume_path =
+    let open Result_syntax in
+    let* control =
+      let path = Irmin_pack.Layout.V5.Volume.control ~root:volume_path in
+      match Io.classify_path path with
+      | `File ->
+          let* cf = Control.open_ ~readonly:true ~path in
+          let+ _ = Control.close cf in
+          Some (Control.payload cf)
+      | `Directory | `Other | `No_such_file_or_directory -> Ok None
```

Comment:
Nitpick: the `classify/open/close/payload` dance could be hidden in a `Control.read`?

Reply:
Good idea to add a `read` function to handle the open/close/payload dance. I'd prefer to keep the `classify` part here since I wouldn't expect the control file module to handle a path that doesn't point to a control file.